### PR TITLE
Checkinout table

### PR DESCRIPTION
**Pull request by allenmichael099**
Made the a table tab to display  bike checkins as a serparate tab from participants and trips. 

From my commit:
- Added function query_checkinout to utils_query_databasee
- Added the checkinout field to data_r in mod_load_data
- called mod_DT_server using data = data_r$checkinout
-  added a tabPanel for "Bike Check In" to app_ui
- moved the assert_flag for anon_locations within run_app

### CODE REVIEW

File: R/app_server.R [hunk 1]
```diff
@@ -126,10 +126,12 @@ app_server <- function( input, output, session ) {
                  dplyr::select(-dplyr::any_of(cols_to_remove_from_participts_table)) %>%
                  data.table::setnames(originalColumnNames,newColumnNames,skip_absent = TRUE)
                )
-    callModule(mod_DT_server, "DT_ui_trips", 
+    callModule(mod_DT_server, "DT_ui_trips",
                data = data_r$trips %>%
                  dplyr::select(-dplyr::any_of(cols_to_remove_from_trips_table)) %>%
                  sf::st_drop_geometry())
+    callModule(mod_DT_server,"DT_ui_checkinout",
+               data = data_r$checkinout)
```
Comment (shankari) [inline R/app_server.R]:
```
>>> for t in scfg["supplementary_tables"]:
...     print("mod_DT_server", "DT_ui_%s" % t["name"], "data = data_r$%s" % t['name'])
...
mod_DT_server DT_ui_Checkinout data = data_r$Checkinout
mod_DT_server DT_ui_PolarBear data = data_r$PolarBear
mod_DT_server DT_ui_SuggestionSys data = data_r$SuggestionSys
mod_DT_server DT_ui_Tier_Sys data = data_r$Tier_Sys
```

File: R/app_ui.R [hunk 2]
```diff
@@ -114,6 +114,12 @@ app_ui <- function(request) {
                         title = "Trips",
                         value = "trips",
                         mod_DT_ui("DT_ui_trips")    
+                      ),
+                      tabPanel(
+                        status = "primary",
+                        title = "Bike Check In",
+                        value = "checkinout",
+                        mod_DT_ui("DT_ui_checkinout")
```
Comment (shankari) [inline R/app_ui.R]:
```
>>> supplementary_tabs = ["tabPanel(status='primary', title=%s, value=%s, mod_DT_ui('DT_ui_%s')" % (t["tab_name"], t["name"], t["name"]) for t in scfg["supplementary_tables"]]
>>> supplementary_tabs
["tabPanel(status='primary', title=Bike Check in, value=Checkinout, mod_DT_ui('DT_ui_Checkinout')", "tabPanel(status='primary', title=Polar Bear, value=PolarBear, mod_DT_ui('DT_ui_PolarBear')", "tabPanel(status='primary', title=Suggestions, value=SuggestionSys, mod_DT_ui('DT_ui_SuggestionSys')", "tabPanel(status='primary', title=Leaderboard, value=Tier_Sys, mod_DT_ui('DT_ui_Tier_Sys')"]
>>> ....expand(supplementary_tabs)
```

File: R/mod_load_data.R [hunk 3]
```diff
@@ -53,6 +53,10 @@ mod_load_data_server <- function(input, output, session, cons){
       summarise_server_calls(., data_r$server_calls)
     message("Finished loading participants")
     
+    message("About to load checkinout")
+    data_r$checkinout <- query_checkinout(cons) %>% data.table::setcolorder(c("user_id"))
+    message("Finished loading checkinout")
+
```
Comment (shankari) [inline R/mod_load_data.R]:
```
>>> for t in scfg["supplementary_tables"]:
...     print("About to load %s" % t["name"])
...     print("data_r$%s = " % (t["name"]), "query_supplementary(cons, t['name'])")
...     print("Finished loading %s" % t["name"])
...
About to load Checkinout
data_r$Checkinout =  query_supplementary(cons, t['name'])
Finished loading Checkinout
About to load PolarBear
data_r$PolarBear =  query_supplementary(cons, t['name'])
Finished loading PolarBear
About to load SuggestionSys
data_r$SuggestionSys =  query_supplementary(cons, t['name'])
Finished loading SuggestionSys
About to load Tier_Sys
data_r$Tier_Sys =  query_supplementary(cons, t['name'])
Finished loading Tier_Sys
```

File: R/utils_query_database.R [hunk 4]
```diff
@@ -38,6 +38,15 @@ query_cleaned_trips <- function(cons) {
...
 
+#' @rdname query
+#' @export
+query_checkinout <- function(cons){
+  # cons <- connect_stage_collections(url = getOption('emdash.mongo_url')) 
+  cons$Checkinout$find('{}') %>%
+    as.data.table() %>%
+    normalise_uuid()
+}
+
```
Comment (shankari) [inline R/utils_query_database.R:49]:
```
query_supplementary <- function(cons, name) {
     cons[name].find('{}')....
}
```

**Review by shankari — COMMENTED** [review]
this review is thinking through what the code changes would look like if the supplementary tables were in the config. I'm working with the config below and using python code.

```
supplementary_tables:
    - Checkinout:
        tab_name: "Bike Check in"
    - PolarBear:
        tab_name: "Polar Bear"
    - SuggestionSys:
        tab_name: "Suggestions"
    - Tier_Sys:
        tab_name : "Leaderboard"
```

File: R/utils_connect_database.R [hunk 5]
```diff
@@ -41,11 +41,17 @@ create_connection <- function(db, collection, url) {
 "COLLECTIONS"
 
 COLLECTIONS <- 
-  c("Stage_Profiles",
+  c("CarbonUsage",
+    "Checkinout",
+    "New_tier",
+    "PolarBear",
```
Comment (shankari) [inline R/utils_connect_database.R:48]: These are all specific to the gamified version that is used in CanBikeCo and should not be in the standard dashboard. Focusing on the supplemental tables,

```
>>> COLLECTIONS=["Stage_timeseries", "Stage_analysis_timeseries", "Stage_profile"]
>>> COLLECTIONS.extend([t["name"] for t in scfg["supplementary_tables"]])
>>> COLLECTIONS
['Stage_timeseries', 'Stage_analysis_timeseries', 'Stage_profile', 'Checkinout', 'PolarBear', 'SuggestionSys', 'Tier_Sys']
```

**Review by shankari — COMMENTED** [review]
One more!

**Comment by shankari** [comment]
Seems like it should all work. Hopefully R has a way to retrieve keys by name (e.g. the equivalent of `data[t["name"]]`). I'm not sure if the `data_r$t["name"]` would work 😄

**Comment by asiripanich** [comment]
> 
> 
> Seems like it should all work. Hopefully R has a way to retrieve keys by name (e.g. the equivalent of `data[t["name"]]`). I'm not sure if the `data_r$t["name"]` would work 😄

Yes, that works in R as well `data_r$t[["name"]]`.

File: R/app_server.R [hunk 6]
```diff
@@ -80,11 +80,18 @@ app_server <- function( input, output, session ) {
   # Tables ------------------------------------------------------------------
   data_esquisse <- reactiveValues(data = data.frame(), name = "data_esquisse")
   
+  # Convert the config's label map to a named vector. 
+  # The names are the original labels, and the vector values are the new labels.
+  named_label_vector <- unlist(getOption('emdash.col_labels_for_participts'))
+  originalColumnNames <- names(named_label_vector)
```
Comment (asiripanich) [inline R/app_server.R]: Let's stick with snake_case for variable naming.

**Comment by asiripanich** [comment]
@allenmichael099 Once I merge #32 to master, could you please pull new commits from master to this branch? Thanks

**Comment by allenmichael099** [comment]
After merging, I ran into an issue related to tidy_cleaned_trips and am currently looking into it. My first error was:
Warning: Error in setorderv: some columns are not in the data.table: data.end_fmt_time

**Comment by allenmichael099** [comment]
So the difference is within query_cleaned_trips: my working version said analysis/cleaned_trip, but after I merged it said analysis confirmed_trip, which returned a null data frame %>% null data table %>% couldn't use setorder. I'm not sure whether or not this has to do with differences between test-data and the actual data set.

There also is an issue just before generate trajectories so that when generate_trajectories is called, tidied_trips is missing 'fmt_time_utc', which gives:
 Warning: Error in eval: object 'fmt_time_utc' not found

**Comment by shankari** [comment]
@allenmichael099 aha! As part of https://github.com/asiripanich/emdash/pull/23
I switched to reading `confirmed_trip` instead of `cleaned_trip`
I also updated the test database
Screen Shot 2021-06-08 at 4 03 34 PM

You need to update your test mongodb and then everything should work 👍 
This will also display `mode_confirm`, `purpose_confirm` etc in the trip table

**Comment by allenmichael099** [comment]
@shankari Got it, thanks! I changed cleaned trip back to confirmed trip and downloaded the new test data. Everything looks like it works now

File: inst/app_config.yml [hunk 7]
```diff
@@ -0,0 +1,85 @@
+# Use this file to set options for the app
```
Comment (asiripanich) [inline inst/app_config.yml]: @allenmichael099 there is already a config-default.yml in inst. Maybe we should merge the two?

Comment (asiripanich) [reply inst/app_config.yml]: Or is there any particular reason why you wanted to put this config file?

**Review by asiripanich — COMMENTED** [review]
Tested! I just have a few comments, but overall the PR looks good to me.

File: R/app_server.R [hunk 8]
```diff
@@ -125,15 +125,45 @@ app_server <- function(input, output, session) {
...
+               data = data_r$trips %>%
+                 dplyr::select(-dplyr::any_of(getOption("emdash.cols_to_remove_trips_table"))) %>%
+                 sf::st_drop_geometry()
     )
+    
+    tableList <- getOption('emdash.supplementary_tables')
+
+    # For each supplementary table, append a new tabPanel and run the server function
+    # that specifies table behavior
+    for (t in tableList){
```
Comment (asiripanich) [inline R/app_server.R:142]: Now that we can have supplementary tables, you may want to look into how we can keep the plot panel above in sync with the table below.

File: inst/app_config.yml (hunk 7 above)
Comment (allenmichael099) [reply inst/app_config.yml]: No particular reason, I just wanted to keep the last two table suggestions somewhere. I'll take it out now.

**Comment by shankari** [comment]
@asiripanich I am going to merge this into a local branch for deployment now

**Comment by shankari** [comment]
This works great barring the sync between the table and the plot. I even enabled the commented out "Polar Bear" table and it worked just fine.

**Comment by shankari** [comment]
Looked at the table with the end user today, and everything is fine except that the timestamp is in milliseconds.
We should really handle that in the dashboard instead of forcing the user to convert manually.

There are two ways in which we can do this:
- convert on the client (perhaps automatically by looking at fields called `ts`)
- store the formatted time on the server

In both cases, we need to decide how to specify the timezone for the conversion

One option is to have the phone send up the formatted time and the timezone, similar to the current background sensed entries. This would require no further customization on the dashboard.

**Comment by shankari** [comment]
Timezones are tricky to work with.
The classic solution for time display is that you store everything as a `ts` and convert at the presentation layer, before displaying to the user.
BUT when you display to the user, **which timezone should you use?**

- no timezone = UTC = time in GMT. This is what AWS etc do, for example. This will be the same no matter where in the world you view it from, which is great for consistency, but requires the user to do mental math to figure out the conversion to their local timezone.
- timezone of the viewer. you can determine this from the computer/browser settings and just convert. This is what most TZ aware code (e.g. Github) does.
- timezone of the checkin/out. So if the user checked out in MDT, show this in MDT even if the viewer is in PDT. That's what we do with the trip start/end. If we want to do this, the fmt time and/or the timezone have to be generated from the app and stored along with the timestamp.

**Comment by shankari** [comment]
This is only really applicable to app-generated supplementary data. Not sure how common a use case that is, and how people are going to want to use it in the future. Let's not overengineer. If @allenmichael099 can get the formatted time in the timezone of the computer along with automatically converting all `ts` columns, that is good enough for now.

If we need to send `fmt_time` from the app, we can migrate the server side data later.

**Pull request closed without merging**